pystac/html/Macros.jinja2

```diff
@@ -0,0 +1,91 @@
+{% macro square(background_color='#E1E1E1', border_color='#9D9D9D') -%}
+    <div style="
+        width: 24px;
+        height: 24px;
+        background-color: {{ background_color }};
+        border: 3px solid {{ border_color }};
+        border-radius: 5px;
+        position: absolute;">
+    </div>
+{%- endmacro %}
+
+
+{% macro extra_fields(parent) -%}
+    {% if parent.extra_fields %}
+        {% for key, value in parent.extra_fields.items() %}
+            <tr><td style="text-align: left;"><strong>{{ key }}:</strong> {{ value }} </td></tr>
+        {% endfor %}
+    {% endif %}
+{%- endmacro %}
+
+
+{% macro links(parent) -%}
+    {% if parent.links|length > 0 %}
+        <details>
+            <summary style="margin-bottom: 10px; margin-top: 10px;">
+                <h4 style="margin-bottom: 0px; display: inline;">Links</h4>
+            </summary>
+            {% for link in parent.links %}
+                {{ link._repr_html_() }}
+            {% endfor %}
+        </details>
+    {% endif %}
+{%- endmacro %}
+
+
+{% macro items(parent) -%}
+    {% if parent.get_items()|list|length > 0 %}
+        <details>
+            <summary style="margin-bottom: 20px;">
+                <h4 style="margin-bottom: 0px; display: inline;">Items</h4>
+            </summary>
+            {% for item in parent.get_items() %}
```

Comment:
A call to `get_items` is going to resolve all Items for this catalog (not recursively) by making a network request to each `"item"` link in the catalog, which could take a pretty long time. I think we probably want to avoid that here and maybe just summarize the item link information instead (i.e. using `get_item_links` and summarizing the title, href, rel type, etc.), even though it does not provide as much detail.

Also, [in `pystac_client.CollectionClient`](https://github.com/stac-utils/pystac-client/blob/47812c6f59e8f81956e62f2149d4becc34e7aae0/pystac_client/collection_client.py#L14), which inherits from `pystac.Collection`, this will result in multiple paginated requests to the API to fetch all Items, which could take a *very* long time. That is more of a concern for that library, but I thought it was worth mentioning (cc: @matthewhanson, @gadomski).

Reply:
I see. I'm torn here: the fact that I can actually click through the items/children is why I wanted to write this in the first place.

Could you please give me an example of loading a STAC catalog over a network / internet, so I could test this out? I only found examples using `Catalog.from_file()` in the documentation.

Reply:
> I see. I'm torn here: the fact that I can actually click through the items/children is why I wanted to write this in the first place.

I agree, I'm torn as well because there is not much information in the links. Could we maybe put a limit on the number of Items that are fetched using something like: `itertools.islice(parent.get_items(), 100)` and give some visual indication that we are only showing the first 100 (or whatever number we choose) items? That way users could get a feel for the Items in the catalog while still limiting the number of requests.

> **edit**: I see that simply passing URL seems to work to some extent, e.g.
> 
> ```python
> cat = pystac.Catalog.from_file('https://cbers-stac-1-0.s3.amazonaws.com/CBERS4/catalog.json')
> ```
> 
> But for example `next(cat.get_children())` hangs forever for me, which I'm not sure if supposed to happen.

I'm not sure what's going on there; that code is running for me in a second or two:

```python
>>> import pystac
>>> cat = pystac.Catalog.from_file('https://cbers-stac-1-0.s3.amazonaws.com/CBERS4/catalog.json')
>>> next(cat.get_children())
<Collection id=CBERS4-MUX>
```

Is the issue persistent for you?

Reply:
> Is the issue persistent for you?

Aha, I think it hangs only when I call `_repr_html_()`, otherwise it works fine. So I now have an example to try :+1: 

> I agree, I'm torn as well because there is not much information in the links. Could we maybe put a limit on the number of Items that are fetched using something like: `itertools.islice(parent.get_items(), 100)` and give some visual indication that we are only showing the first 100 (or whatever number we choose) items?

Yes, I like that, I will try it out. 

I'm a bit worried that my "hangs forever" issue is probably because the templates get constructed by fetching all items/children immediately and recursively. If that's the case, then even a limit on the number of fetched items could be too simple. 

I will look into it, looking into some lazy loading options (e.g. there [seem to be some HTML mechanisms](https://developer.mozilla.org/en-US/docs/Web/Performance/Lazy_loading)

Reply:
Here is what I tried

# Loading only the first child/item
For the `CBERS` dataset, this took ~2.2s to render. Note that it's fairly nested (4 levels of nesting to get to an item). Two seconds seems reasonable to me, especially since that only happens the first time.

Displaying any additional items increases the time exponentially and is IMO not feasible. Loading two children took ~10s.

### Example
```python
import pystac
cat = pystac.Catalog.from_file('https://cbers-stac-1-0.s3.amazonaws.com/CBERS4/catalog.json')
```

### Preview
![image](https://user-images.githubusercontent.com/18722560/154703651-99bd9708-eb91-4fe0-873a-8471cb7cdfe7.png)

# Lazy loading
I looked into the pure HTML/CSS options for lazy loading, but didn't find anything useful.
- HTML: `loading="lazy"` only seems to be available for `img` and `iframe`
- CSS: `content-visibility` still loads the data immediately, only controls visibility

Reply:
I don't see a better solution for now. Showing only the first entry is a reasonable solution IMO.

Reply:
Yeah, I agree this seems like the best solution for now.